tag-historian/nitaghistorian.yml

```diff
@@ -228,26 +173,40 @@ definitions:
         format: iso-date-time
         example: '2018-09-04T18:45:08Z'
         default: '0001-01-01T00:00:00Z'
-      skip:
-        description: Limits the returned historical values to those after the given number of values.
-          Use with the take property to implement pagination. Only applies when not specifying a
-          decimation.
-        type: integer
-        minimum: 0
-        default: 0
       take:
         description: Limits the returned historical values to the specified number of values. Use
-          with the skip property to implement pagination. Only applies when not specifying a decimation.
+          with the continuationToken property to implement pagination. Only applies when not specifying a decimation.
         type: integer
         minimum: 0
         default: 0
         example: 1
+      continuationToken:
```

Comment:
I would say a single continuationToken with multiple tag paths does not make a lot of sense...

We should probably change "paths" to only support a single tag "path".

Reply:
The LabVIEW API currently has a multi-read history VI. We'll need to determine if this means we would remove that VI entirely, simulate it client-side by making a request per tag, or continue supporting it in the HTTP API. @prestwick, what do you think about potentially dropping support for the Multi-Read History?

Reply:
Also, in the current API, skip and take doesn't apply to decimated reads. Will that still be true with the continuation token? Both the web UI and dashboards will want to be able to read decimated history for multiple tags for display on a single graph.

Reply:
> I would say a single continuationToken with multiple tag paths does not make a lot of sense...

It seems like the service could process the tag paths in order, and assume that whatever tag comes after the one referenced in the current continuation token is the next one to process.

Reply:
Any process on this one? We can encode the necessary information for each requested tag path in the continuation token. It just seems quite weird and overly complex to iterate over multiple tag paths.

_It seems like the service could process the tag paths in order, and assume that whatever tag comes after the one referenced in the current continuation token is the next one to process._

Yeah, but isn't the API supposed to return multiple results at once, meaning if you ask for tag "A" and "B", the service returns the first 100 values of "A" and first 100 values of "B" in the first response?

Reply:
We're most concerned about being able to supply multiple tag paths for decimated queries, as users could have dashboards with tiles containing many tags. Chrome and Firefox limit users to 6 concurrent requests per host, so being able to specify multiple tags per request is useful in that context. The service could process each tag in the query in parallel.

We're less sure about undecimated queries. One option could be to change the behavior for undecimated queries such that the service queries the tags one at a time, rather than all at once, as we agree that the current behavior is strange. Continuation tokens would be more straightforward with that model. Another option would be to split the undecimated query stuff out into a separate route, and possibly make that route only support querying for one tag at a time.

> Also, in the current API, skip and take doesn't apply to decimated reads. Will that still be true with the continuation token?

To reiterate Paul's question above: Will continuation tokens apply to decimated queries? It seems like they shouldn't.

Reply:
_Will continuation tokens apply to decimated queries?_
They would not apply to decimated queries.

_The service could process each tag in the query in parallel._
The client can just send the requests in parallel. This also makes it easier to apply throttling.

_ Chrome and Firefox limit users to 6 concurrent requests per host_
6 concurrent requests should really be enough if our queries are in the ms. And it is really not relevant with websocket and http2 anyways.

_Another option would be to split the undecimated query stuff out into a separate route_
Yeah, I think they are different enough that it would make sense... Otherwise the API needs a lot of documentation, explaining for decimated queries you can specify fields a,b,c and for undecimated queries fields a,d,e...

Reply:
> The client can just send the requests in parallel. And we can basically start displaying tiles when the first request gets returned and not when the slowest query finally finished.
This also makes it easier to apply throttling.

There are a lot of clients which would be impacted by such a change. This would break dashboard builder, and represent additional breakages to the G API and tag viewer. I would prefer the complexity of parallellizing queries to be in the service, rather than push that complexity to each of the clients of the API. If you still feel strongly about breaking the API in this way, I think we should have a broader discussion at the spec review meeting with all of the stakeholders who would be impacted.

> Yeah, I think they are different enough that it would make sense...

Carson tells me that dashboard builder doesn't do undecimated queries. I don't think tag viewer does either. The G API does, but I think we could handle that by just querying a different route in that case. I'll push changes to detail what the API might look like if the undecimated query stuff was split out into its own route.

You and the other SLC folks should also feel free to push to this branch if there's something you'd like to change, also.

Reply:
Another reason to continue to support specifying multiple tags is that this route is also used to export tag history to CSV files. If a user wants to export the history data of three tags to CSV, @prestwick's feeling is that they will want that data in one CSV file, not three. We don't want users, or our web UI clients, to have to make three requests for each tag and then stitch the CSV files together.

Reply:
@spanglerco brought up the point that /v1/tags/query-history seems like a suitable path for *undecimated* queries, so it's actually the decimated queries that would be split out into a new route... I think I'm going to go with /v1/tags/query-decimated-history for now.

Reply:
_There are a lot of clients which would be impacted by such a change._
everything what we discuss here is already somewhat breaking:
- removing skip
- adding continuation token support
- removing metadata
- removing complete routes
- etc...

I would say that if we already make these changes, then everything should be on the table to make this API more performant, concise and easier to understand.

If we don't want to break anybody than we have to keep the old API on SLS... and we just call this API /v2/. And maybe systemlink cloud will never implement /v1/

Reply:
_Another reason to continue to support specifying multiple tags is that this route is also used to export tag history to CSV files._

Doesn't need to be. If we want to support larger CSV downloads, we could provide a separate download API which is even async. And this can be used if clients want to perform data retention. what do you think?

Reply:
> I would say that if we already make these changes, then everything should be on the table to make this API more performant, concise and easier to understand.

But there's a difference between an API break and changing the supported feature set. In this instance, Phil is referring to the impact of a feature loss on clients that are using that feature in a visible way.

> If we want to support larger CSV downloads, we could provide a separate download API which is even async.

No need to make it async when a CSV file can be streamed as it's generated. But the same use case applies to LabVIEW applications as well. What would be the difference between a download API and query-data? One would support multiple paths but not continuation token and the other would be single path with continuation token? I don't really see the benefit when the continuation token can be used to iterate over each tag path in turn.